ZOOKEEPER-1112: Add (Cyrus) SASL authentication support to C client library

ztzg opened the pull request:
This is a "respin" of https://github.com/apache/zookeeper/pull/1054, which I withdrew due to some annoying shortcomings.

This changeset allows C clients to use SASL to authenticate with the ZooKeeper server.  It is loosely based on patches #1 and #2 by Tom Klonikowski, at https://reviews.apache.org/r/2252/, but the result has been extensively reworked to follow the semantics of the Java client:

  * No SASL operations are exposed through the API;

  * The configuration is provided, and stored, at "handle init time";

  * SASL authentication is automatically performed after each (re)connect.

It introduces an optional dependency on the Cyrus SASL library, which can either be autodetected (default) or configured using the `--without-sasl`/`--with-sasl[=DIR]` flags, or -DWITH_CYRUS_SASL for CMake/Windows.

`TestServerRequireClientSASLAuth.cc` has been renamed to `TestSASLAuth.cc`, and a test has been added which successfully (re)authenticates using the `DIGEST-MD5` mechanism.  The code has also been used to successfully authenticate clients via `GSSAPI`/Kerberos.

This commit also adds SASL support to the `cli.c` client.

Co-authored-by: Tom Klonikowski <klonik_t@informatik.haw-hamburg.de>

ztzg commented:
Pinging @eolivelli @phunt @anmolnar and @hanm, which were mentioned on the earlier https://github.com/apache/zookeeper/pull/1054, and @enixon, who seems to be willing to review C components :)

ztzg commented:
Okay.  The refreshed version I just pushed:

1. has been rebased on top of the OpenSSL support;
2. is partially aligned with the (still unmerged) OpenSSL detection patches;
3. builds and runs under Windows;
4. suffers from a regression in the test suite (but interactive use works well).

TODO items:

1. Fix regression;
2. Finish aligning with the OpenSSL detection, license & docker patches.

(Cc: @symat)

ztzg commented:
The last refresh:

  * Has been rebased on top of https://github.com/apache/zookeeper/pull/1159 (includes SSL support and OpenSSL autodetection);
  * Is "fully aligned" with https://github.com/apache/zookeeper/pull/1159 (Maven, Docker, Autoconf and CMake, as well as license description);
  * Builds and can SASL-authenticate under Windows;
  * Adds a working `DIGEST-MD5` SASL test;
  * Enables SASL authentication in `cli_st`/`cli_mt` (`DIGEST-MD5`/`GSSAPI`; try `cli --help` for examples).

This PR has no known issues and/or known outstanding points, so this message also is a review/merge request.  Pinging @eolivelli, @phunt, @anmolnar and @hanm, which were mentioned on the earlier #1054; @enixon, who seems to be willing to review C components; as well as @symat, whose SSL efforts served as a guide.

symat commented:
great stuff, thanks!
I will try to review + test on a real kerberized cluster today / tomorrow.

ztzg commented:
@symat: Thanks! :)

symat commented on zookeeper-client/zookeeper-client-c/src/cli.c:
```diff
@@ -840,10 +877,26 @@ int main(int argc, char **argv) {
                 argv[0]);
+#ifdef HAVE_CYRUS_SASL_H
+        fprintf(stderr,
+                "SASL EXAMPLES:\n"
```
please also mention here, that SSL and SASL are not working together at the moment. AFAICS in the implementation, if the user defines both SSL and SASL parameters, then only the SASL will be used. (I am not sure what happens, if the SASL fails to authenticate... with in this case fallback to SSL?)

ztzg replied:
Done.  (But see the note above above, about `zookeeper_init_with_options`.)

ztzg commented on zookeeper-client/zookeeper-client-c/Makefile.am:
```diff
@@ -59,6 +65,7 @@ libzkmt_la_LIBADD = -lm $(CLOCK_GETTIME_LIBS)
 
 lib_LTLIBRARIES += libzookeeper_mt.la
 libzookeeper_mt_la_SOURCES =
+libzookeeper_mt_la_CFLAGS = -DTHREADED
```
This is a leftover from the various rebases the patch has gone through.  Unrelated to SASL, and unnecessary as `libzookeeper_mt` is made of already-compiled objects. Removed.

ztzg commented on zookeeper-client/zookeeper-client-c/include/zookeeper.h:
```diff
@@ -575,6 +579,93 @@ ZOOAPI zhandle_t *zookeeper_init2(const char *host, watcher_fn fn,
   int recv_timeout, const clientid_t *clientid, void *context, int flags,
   log_callback_fn log_callback);
 
+#ifdef HAVE_CYRUS_SASL_H
```
Okay, good to know!

A related problem is that the SSL support adds `zookeeper_init_ssl`, whereas SASL adds `zookeeper_init_sasl`.  That obviously does not compose.  I suppose we'll have to introduce something like `zookeeper_init_with_options(..., zookeeper_options_t *)` at some point—but one step at a time.

ztzg commented on zookeeper-client/zookeeper-client-c/src/cli.c:
```diff
@@ -840,10 +877,26 @@ int main(int argc, char **argv) {
+                "-n, --mechlist <mechlist>      Comma separated list of SASL mechanisms\n"
+                "-u, --user <user>              SASL user name\n"
+                "-l, --realm <realm>            SASL realm\n"
+                "-p, --password-file <file>     File containing the password for SASL\n"
```
I have added a note regarding `--password-file`, and clarified the use of parameters in general (they depend on the chosen SASL mechanism; missing bits are obtained via callbacks).

ztzg commented on zookeeper-client/zookeeper-client-c/src/cli.c:
```diff
@@ -876,18 +929,43 @@ int main(int argc, char **argv) {
+        zoo_sasl_params_t sasl_params = { 0 };
+
+        if (sasl_client_init(NULL) != SASL_OK) {
+            return 1;
```
This is the status code which is returned to the OS.  `cli.c` returns a mixture of 1, 2 and sometimes `errno` (?!) on error.  It would be good to rationalize this, but I would suggest doing it in one go (and perhaps just using `EXIT_FAILURE`).

I have added some diagnostics, however, as there is no good reason to "just exit."

ztzg commented on zookeeper-client/zookeeper-client-c/tests/TestSASLAuth.cc:
```diff
@@ -0,0 +1,188 @@
+        stopServer();
+    }
+#endif /* HAVE_CYRUS_SASL_H */
+};
```
Done.

ztzg commented:
Hi @symat,

Thank you for the most excellent review and testing!

> Overall it looks good! :) The code is really nice and it is great that you also added tests / license / build options / documentation. I had a few minor comments / questions, please check them if you can.

I believe I have addressed everything.  (Things could undoubtedly be polished some more, but I don't know of any important outstanding issues at this point.)

> I also tested the patch by:
> 
> * compile it on Centos 7.5
> * execute all the C unit tests
> * use both `cli_mt` and `cli_st` to connect to a ZooKeeper using an external MIT Kerberos for authentication

Great to read that you did not encounter issues, as your setup is quite a bit different from mine.

symat commented:
Thanks for all the changes, they all look OK for me. I quickly re-built everything and was able to test it with Kerberos. However, some unit tests failed for me this time:
- TestReadOnlyClient -> testReadOnlyWithSSL
- TestReadOnlyClient -> testReadOnly
(I don't know how I didn't notice it before, I guess this problem is unrelated to your latest changes)

I also tested the `-r` / `--readonly` option with the CLI and it allowed me to create / set znodes, so I think something seems to be wrong here and it is not only a test issue. (it also doesn't seems to work when I tried read-only mode without SASL... however, I haven't tested this without your patch)

I remember that when I did the SSL support, I had to touch this part. The C API is using some 4-letter-word command on a separate socket to verify something about the read-only mode, see here: https://github.com/apache/zookeeper/blob/a80f9b291f2852b8dbe3b83bf2b5b4f0306424f5/zookeeper-client/zookeeper-client-c/src/zookeeper.c#L2191
I had to initiate the SSL connection for this extra socket too. Maybe you are hitting this part?

(It might also worth to add an extra `testReadOnlyWithSASL` testcase, when it is fixed)

ztzg commented:
hi @symat,

`-r`/`--read-only` sets `ZOO_READONLY`, which really is misnamed.  It should be `ZOO_CAN_BE_READONLY`, to match the Java parameter:

     * @param canBeReadOnly
     *                whether the connection is allowed to go to read-only
     *                mode in case of partitioning

You can try the same thing with the Java "cli", and create/modify nodes nilly-willy:

    zkCli -r -server localhost:2181

The SASL patches just perform a handshake on top of the existing connection, so "4lw" access and other connection parameters should be irrelevant—besides reverse DNS resolution, which is indeed needed for "proper" `GSSAPI`/Kerberos auth.

ztzg commented:
> Thanks for all the changes, they all look OK for me. I quickly re-built everything and was able to test it with Kerberos. However, some unit tests failed for me this time:
> 
> * TestReadOnlyClient -> testReadOnlyWithSSL
> * TestReadOnlyClient -> testReadOnly
>   (I don't know how I didn't notice it before, I guess this problem is unrelated to your latest changes)

I tried in a few environments, but could not reproduce this.  Could it be that you had a leftover ZooKeeper running?  Or that you were using a VM which might cause ZK to take more than 4s to start up?  There is a "hack" in `zkServer.sh` for `startCleanReadOnly` and `startRequireSASLAuth`:

       sleep 4
       success=true

> I also tested the `-r` / `--readonly` option with the CLI and it allowed me to create / set znodes, so I think something seems to be wrong here and it is not only a test issue. (it also doesn't seems to work when I tried read-only mode without SASL... however, I haven't tested this without your patch)

See https://github.com/apache/zookeeper/pull/1134#issuecomment-567134796, above.
 
> (It might also worth to add an extra `testReadOnlyWithSASL` testcase, when it is fixed)

Done & pushed.

symat commented:
OK, I think you are right. (I remember checking the 'read-only' feature before, jut forgot the thing... the parameter name is a bit misleading indeed)

Anyway, thanks for looking into it and also adding the extra test!

Something must be with my environment then. (I am running it on mac from docker, maybe it is some timing issue / memory limitation.) I just re-executed the tests and now I got different failures. But I checked the apache jenkins jobs and all the tests were executed / green there, so it is good enough for me :)

ztzg commented:
> Something must be with my environment then. (I am running it on mac from docker, maybe it is some timing issue / memory limitation.) I just re-executed the tests and now I got different failures. But I checked the apache jenkins jobs and all the tests were executed / green there, so it is good enough for me :)

Right.  I suspect something is going on between Docker & the test version of `zkServer.sh`.  I have tested in an Ubuntu 16.04 `nspawn` container, and on a NixOS 19.09 host, and did not see any such errors.  (It would be good to "robustify" `zkServer.sh` at some point; that `sleep 4` isn't the most reasonable thing in the world :)

Sorry for the SPAM, but @eolivelli, @phunt, @anmolnar, @hanm, @enixon: FYI, Máté was able to SASL-auth with a Kerberos-enabled cluster from the C client.  I don't have any outstanding issues, now that this has been rebased and aligned on C client SSL support.  What do you think?

ztzg commented:
(Originally a reply to this message of Enrico Olivelli on the mailing
list:
https://mail-archives.apache.org/mod_mbox/zookeeper-dev/201912.mbox/%3CCACcefgfEx2VHaJiy9YYW4WZO9TQ2-pjNyDOy8Ew%2BUWFERQnGUA%40mail.gmail.com%3E )


Hi Enrico, all,

> We have the last patch that is waiting for review:
> https://github.com/apache/zookeeper/pull/1134/
>
> https://issues.apache.org/jira/browse/ZOOKEEPER-1112 Add support for C
> client for SASL authentication
>
> It is a very big patch on the C client and it needs careful review and it
> is introducing a lot of changes (many LOCs).

I agree that the patch is substantial.  I tried, however, to minimize
the impact on the existing client code in order to avoid regressions
when SASL support is compiled out or disabled.

I would encourage reviewers to question my choices, either on this list
or on the "pull request"—but perhaps a small "review guide" can help in
the meantime:


M       zookeeper-client/zookeeper-client-c/src/zookeeper.c

The most critical change I was not able to avoid lies in 'zookeeper.c':
I was forced to split the "bottom half" of the session establishment
logic to a 'finalize_session_establishment' procedure.  That code still
is invoked "synchronously" from the same location when SASL is disabled
or inactive—but emulating the Java client causes that finalization to
happen after (asynchronous) SASL negociation.

The code in 'zookeeper.c' now also tests 'is_sasl_auth_in_progress' from
a few locations.  This is analoguous to what happens in the Java client.
These tests should (hopefully!) be relatively innocuous as the answer is
trivially false unless SASL has been compiled in *and* requested.


The rest of the changes are fairly straightforward:

M       zookeeper-client/zookeeper-client-c/src/zk_adaptor.h

Adds an optional 'zoo_sasl_client_t *' to 'zhandle'.  This changes the
size of the struct, but should otherwise have no effect unless SASL is
requested.

M       zookeeper-client/zookeeper-client-c/include/proto.h

Just adds a missing #define.

M       zookeeper-client/zookeeper-client-c/include/zookeeper.h

Adds new definitions, but does not modify existing ones.

A       zookeeper-client/zookeeper-client-c/src/zk_sasl.c
A       zookeeper-client/zookeeper-client-c/src/zk_sasl.h

New files, contain the bulk of the SASL logic.  These files are not even
compiled in when SASL support is disabled.

M       zookeeper-client/zookeeper-client-c/src/cli.c

Useful for interactive testing of C SASL auth, or as a simple example,
but not expected to impact "production" deployments.

A       zookeeper-client/zookeeper-client-c/tests/TestSASLAuth.cc

Tests added (file used to be called TestServerRequireClientSASLAuth.cc)

M       README_packaging.md
M       dev/docker/Dockerfile
M       pom.xml
A       tools/cmake/Modules/FindCyrusSASL.cmake
M       zookeeper-client/zookeeper-client-c/CMakeLists.txt
M       zookeeper-client/zookeeper-client-c/LICENSE
M       zookeeper-client/zookeeper-client-c/Makefile.am
M       zookeeper-client/zookeeper-client-c/README
M       zookeeper-client/zookeeper-client-c/configure.ac
M       zookeeper-client/zookeeper-client-c/pom.xml
M       zookeeper-client/zookeeper-client-c/tests/zkServer.sh

Misc. configuration/detection logic, documentation updates, LICENCE
updates, notes, etc.


> Personally I am not able to do a real review of C code and it would need
> another committer to be accepted.
>
> I am sorry, I think it will be a great enhancement and the community will
> love it (and Perl/Python bindings will benefit as well), but IMHO it can
> wait 3.6.1 or 3.7.0
> We are not in a hurry.

Note that I already have working SASL updates to the Perl client in my
pipe :) They still need a round of cleanups, but are otherwise ready to
go.

Waiting for 3.6.1 would not be too bad.  But we would really like to see
this land in 3.6.x to avoid carrying such an "unofficial patch" in
production, where it would only get reviews limited by a singular point
of view.

> So if no reviewer chimes in within a couple of days I will postpone that
> issue to the next release.
>
> I am continuing to validate branch-3.6 before sending a formal VOTE
>
> Cheers and season greetings to every one !
> Enrico

Cheers, and season greetings!

Damien Diederen

ztzg commented:
Hi @eolivelli, @anmolnar, @nkalmar, @phunt:

I understand that you are all busy, but am wondering: is there anything else I can do to facilitate reviewing this PR?

I don't know if the ["review guide"](https://github.com/apache/zookeeper/pull/1134#issuecomment-569935984) I posted above is useful.  I don't know if you do your reviews using the GitHub interface, or something else; in any case, let me know if you would prefer me to re-split the patches along the lines of the "guide."

What do you think?

nkalmar commented:
I talked with Mate. Patch is committed to master (3.7.0). Thanks @ztzg !

ztzg commented:
@nkalmar, @symat: Excellent!  Thank you for the reviews; this was a big chunk of code to ingest in one go.  I'm planning to hang around and help iron out the kinks, if any.

Cheers, -D

cuibo01 commented:
> Overall it looks good! :) The code is really nice and it is great that you also added tests / license / build options / documentation. I had a few minor comments / questions, please check them if you can.
> 
> I also tested the patch by:
> 
> * compile it on Centos 7.5
> * execute all the C unit tests
> * use both `cli_mt` and `cli_st` to connect to a ZooKeeper using an external MIT Kerberos for authentication

hi @symat can u help me, how use kerberos and zcli c

symat commented:
Hi @cuibo01 !

I'm not sure about jacc.conf , I never used that. I only use the C client very rarely. You can use a bit nicer CLI parameters for more recent ZK versions. E.g. with my zookeeper I usually do a "kinit", then using the C CLI with parameters, like:
```
./cli_mt -h <zk-server-host>:2181 --mechlist GSSAPI --user <kerberos principal username> --realm <kerberos principal realm> --server-fqdn <kerberos principal host> --debug
```

Although I don't really remember from which ZK version we are supporting this CLI parameter format...

If you need more info, then please use the zookeeper "user" mailing list (user@zookeeper.apache.org, see https://zookeeper.apache.org/lists.html). There is a good chance that more people in the ZooKeeper community already did this configuration and can help you.

cuibo01 commented:
> Hi @cuibo01 !
> 
> I'm not sure about jacc.conf , I never used that. I only use the C client very rarely. You can use a bit nicer CLI parameters for more recent ZK versions. E.g. with my zookeeper I usually do a "kinit", then using the C CLI with parameters, like:
> 
> ```
> ./cli_mt -h <zk-server-host>:2181 --mechlist GSSAPI --user <kerberos principal username> --realm <kerberos principal realm> --server-fqdn <kerberos principal host> --debug
> ```
> 
> Although I don't really remember from which ZK version we are supporting this CLI parameter format...
> 
> If you need more info, then please use the zookeeper "user" mailing list ([user@zookeeper.apache.org](mailto:user@zookeeper.apache.org), see https://zookeeper.apache.org/lists.html). There is a good chance that more people in the ZooKeeper community already did this configuration and can help you.

thx @symat 
if principal is user/host1@CC.COM:
--user is user
--realm is CC.COM
--server-fqdn is host1
right?

symat commented:
yes

cuibo01 commented:
@symat thank you very much
